EventGenerator/src/LeadingLogGenerator_tool.cc

```diff
@@ -0,0 +1,89 @@
+#include "art/Utilities/ToolMacros.h"
+
+#include "CLHEP/Random/RandPoissonQ.h"
+#include "CLHEP/Random/RandGeneral.h"
+
+#include "Offline/EventGenerator/inc/ParticleGeneratorTool.hh"
+
+#include "Offline/DataProducts/inc/PDGCode.hh"
+#include "Offline/MCDataProducts/inc/GenId.hh"
+#include "Offline/Mu2eUtilities/inc/RandomUnitSphere.hh"
+#include "Offline/Mu2eUtilities/inc/BinnedSpectrum.hh"
+#include "Offline/GlobalConstantsService/inc/GlobalConstantsHandle.hh"
+#include "Offline/GlobalConstantsService/inc/ParticleDataTable.hh"
+#include "Offline/GlobalConstantsService/inc/PhysicsParams.hh"
+#include "fhiclcpp/types/DelegatedParameter.h"
+
+namespace mu2e {
+  class LeadingLogGenerator : public ParticleGeneratorTool {
+  public:
+    struct PhysConfig {
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+      fhicl::DelegatedParameter spectrum{Name("spectrum"), Comment("Parameters for BinnedSpectrum)")};
+      fhicl::Atom<int> pdgId{Name("pdgId"),Comment("pdg id of daughter particle")};
+    };
+    typedef art::ToolConfigTable<PhysConfig> Parameters;
+
+    explicit LeadingLogGenerator(Parameters const& conf) :
+      _pdgId(conf().pdgId()), 
+      _mass(GlobalConstantsHandle<ParticleDataTable>()->particle(_pdgId).ref().mass().value()),
+      _spectrum(BinnedSpectrum(conf().spectrum.get<fhicl::ParameterSet>()))
+    {
+      pid = static_cast<PDGCode::type>(_pdgId);
+    }
+
+    std::vector<ParticleGeneratorTool::Kinematic> generate() override;
+    void generate(std::unique_ptr<GenParticleCollection>& out, const IO::StoppedParticleF& stop) override;
+
+    void finishInitialization(art::RandomNumberGenerator::base_engine_t& eng, const std::string& material) override {
+      _randomUnitSphere = new RandomUnitSphere(eng);
+      _randSpectrum = new CLHEP::RandGeneral(eng, _spectrum.getPDF(), _spectrum.getNbins());
+    }
+
+  private:
+    int _pdgId;
+    ProcessCode _process;
+    double _mass;
+
+    BinnedSpectrum    _spectrum;
+
+    RandomUnitSphere*   _randomUnitSphere;
+    CLHEP::RandGeneral* _randSpectrum;
+    double _endPointEnergy;
+    
+    PDGCode::type pid;
+  };
+
+
+  std::vector<ParticleGeneratorTool::Kinematic> LeadingLogGenerator::generate() {
+    std::vector<ParticleGeneratorTool::Kinematic>  res;
+
+    double energy = _spectrum.sample(_randSpectrum->fire());
+
+    const double p = energy * sqrt(1 - std::pow(_mass/energy,2));
+    CLHEP::Hep3Vector p3 = _randomUnitSphere->fire(p);
```

Comment:
I know that everyone does this but it's dangerous.  gcc happens to know that you you mean x*x not exp(2.0*log(x)), which is much, much slower to evaluate.  I don't know about clang or Intel or other compilers we might use down the road.  Look at cetlib/pow.h which lets you write  cetlib::square(x); for an arbitrary integer power there is cetlib::pow<N>(x).  Also, evaluating the difference of squares the way you do it is susceptible to precision issues.  In real Mu2e cases, we won't have precision issues in this line.  The better way to write a difference of squares is (a-b)*(a+b).    And pow.h has a helper function template for that too.

Reply:
Yes, (a-b)*(a+b) is the best here for numeric precision.  For the record, in places where one needs to square a value, there is std::pow(x,2).  I would not use third-party libs since there  is an std equvalent.